### PR TITLE
Partial (and temporary) fix for age restricted videos

### LINKED ISSUE #2189: [Bug] Age Restricted Videos do not play
Videos that have been marked age-restricted don't play
Any video with age-restriction doesn't work. All you can see is the thumbnail. I've tested this on all public instances.

1. Go to any age-restricted video
2. See error

Even the play button doesn't work
image

**Additional context**
 - Browser: Latest Windows version of Chrome
 - OS: Win10 x64

### PR DESCRIPTION
**Pull request by syeopite**
Fixes #2189

This is a temporarily fix for age restricted videos based on what many other projects such as yt-dlp have implemented. However, this only fixes a large portion of age restricted videos but not all of them. Furthermore, it seems like this fix also either exposes or creates issues regarding  captions on a subset of age gated videos. 

If this PR is still compatible with the bounty, I'll like to claim it. Please don't use my previous address though.

### CODE REVIEW

**Comment by unixfox** [comment]
> Furthermore, it seems like this fix also either exposes or creates issues regarding captions on a subset of age gated videos.

This is still better than not being able to load the video at all though.

I'm testing this PR on YewTube.

**Comment by unixfox** [comment]
I tested with this age-restricted video: https://yewtu.be/watch?v=fzhHXt7NKUs. I know you told us that captions could break, but I didn't think the video wouldn't load at all.

```
Title: `Missing JSON attribute: simpleText
  parsing CaptionName at line 1, column 350
  parsing Caption#name at line 1, column 343 (JSON::SerializableError)`
```

I think we should do something about this error, skipping the ability to try fetching the captions?

**Comment by syeopite** [comment]
Theoretically it should be a fairly simple fix. "Normal" captions are sent in this format:
`{"languageCode" => "fr", "languageName" => {"simpleText" => "French"}}`

But for some age restricted videos it becomes 

`{"languageCode" => "fr", "languageName" => {"runs" => [{"text" => "French"}]}}`

The code for parsing it is in [videos.cr](https://github.com/iv-org/invidious/blob/master/src/invidious/videos.cr)

https://github.com/iv-org/invidious/blob/5ba9a1f87d876464f5e7cf72777797c9772aebca/src/invidious/videos.cr#L785-L789

I'm just not too familiar with how `JSON::Serializable` works though.

**Comment by pukkandan** [comment]
Whatever age-gate video doesn't work with this method didn't use to work previously either. See https://github.com/yt-dlp/yt-dlp/pull/328#issuecomment-864117168

File: src/invidious/videos.cr [hunk 1]
```diff
@@ -989,9 +989,15 @@ def fetch_video(id, region)
...
   if info["reason"]?
-    embed_page = YT_POOL.client &.get("/embed/#{id}").body
-    sts = embed_page.match(/"sts"\s*:\s*(?<sts>\d+)/).try &.["sts"]? || ""
-    embed_info = HTTP::Params.parse(YT_POOL.client &.get("/get_video_info?html5=1&video_id=#{id}&eurl=https://youtube.googleapis.com/v/#{id}&gl=US&hl=en&sts=#{sts}").body)
+    required_parameters = URI::Params.new({
+      "video_id" => [id],
+      "eurl"     => ["https://youtube.googleapis.com/v/#{id}"],
+      "html5"    => ["1"],
+      "c"        => ["TVHTML5"],
+      "cver"     => ["6.20180913"],
```
Comment (SamantazFox) [inline src/invidious/videos.cr:997]: I don't think that putting each parameter into an array is required.

Comment (syeopite) [reply src/invidious/videos.cr:997]: Yeah that would be cleaner but unfortunately, the [`URI::Params.new`](https://crystal-lang.org/api/1.0.0/URI/Params.html#new(raw_params:Hash(String,Array(String)))-class-method) constructor requires a type of `Hash(String, Array(String)))`. I'll go ahead and switch to [`URI::Params.encode`](https://crystal-lang.org/api/1.0.0/URI/Params.html#encode(hash:Hash(String,String%7CArray(String)))-class-method) which does allow for just.

**Comment by SamantazFox** [comment]
@syeopite can you please wrap the comment above the HTTP request?

**Comment by SamantazFox** [comment]
I'm going to deploy that on https://test.invidious.io tomorrow. Currently #2196 is live there, then I'll have to fix #2193, then I have to backup the DB and try #1678, and finally I can go with this PR :)

**Comment by syeopite** [comment]
I could go ahead and fix up #2193 if you'd like me to. I've been cleaning up #2059 for an initial merge anyways so adding the required sanitization to that shouldn't be too difficult.

**Comment by SamantazFox** [comment]
> I could go ahead and fix up #2193 if you'd like me to. I've been cleaning up #2059 for an initial merge anyways so adding the required sanitization to that shouldn't be too difficult.

If you want to fix #2193, don't touch to `src/invidious/views/components/item.ecr`, as it will be fixed with #2196!

**Comment by unixfox** [comment]
We really should look into fixing this bug:

> ```
> Title: `Missing JSON attribute: simpleText
>   parsing CaptionName at line 1, column 350
>   parsing Caption#name at line 1, column 343 (JSON::SerializableError)`
> ```

**Comment by syeopite** [comment]
There we go, the caption parsing should be fixed now.

**Comment by unixfox** [comment]
> There we go, the caption parsing should be fixed now.

Yes, it is working perfectly, thank you! I just applied your patch.

We will wait a couple of days for testing purposes and if we don't encounter any issue this is probably good to be merged.

**Comment by SamantazFox** [comment]
> We will wait a couple of days for testing purposes and if we don't encounter any issue this is probably good to be merged.

@unixfox You've deployed in on yewtu.be?

**Comment by unixfox** [comment]
> > We will wait a couple of days for testing purposes and if we don't encounter any issue this is probably good to be merged.
> 
> @unixfox You've deployed in on yewtu.be?

Yes that's why I did set this PR to "in-testing".

**Comment by SamantazFox** [comment]
> Yes that's why I did set this PR to "in-testing".

Ok, nice :)

**Comment by neurodiverseEsoteric** [comment]
test.invidious.io still doesn't play: https://www.youtube.com/watch?v=KgmoMO66uPg

**Comment by syeopite** [comment]
Because this PR still isn't deployed on https://test.invidious.io. But it is running on yewtu.be, and the video works perfectly there.

**Comment by neurodiverseEsoteric** [comment]
Ah ok, but it's not possible to download from yewtu.be

**Comment by unixfox** [comment]
@syeopite Do you think you could trigger this bypass age restriction only when we detect it? With something like this:
```
if info["reason"]?.try &.as_s.includes?("confirm your age")
```

We don't exactly know if this can bypass the same other restricted videos (not age restricted videos) like the old method with the embed.

**Comment by syeopite** [comment]
Are there even any videos that are restricted besides age gated ones? The only other I know of would be the videos that are geoblocked but those are handled elsewhere.

**Comment by unixfox** [comment]
> Are there even any videos that are restricted besides age gated ones? The only other I know of would be the videos that are geoblocked but those are handled elsewhere.

I don't know either, but that's the issue, we can't test that this won't break any ability to watch restricted videos (not age or geo restricted) because we have no dataset. That's why it's safer to keep the old method as a last measure of trying to bypass a restriction.

**Comment by syeopite** [comment]
Fair enough. The old method has been readded as a fallback in 57bb8c610a5489a7159a29004fdd56ec8bff50e5

**Comment by syeopite** [comment]
Merge status?

**Comment by unixfox** [comment]
For me it's safe to merge! It is working great.

File: src/invidious/videos.cr [hunk 2]
```diff
@@ -782,18 +786,19 @@ struct Video
   end
 end
 
-struct CaptionName
-  include JSON::Serializable
+class Caption
```
Comment (SamantazFox) [inline src/invidious/videos.cr:789]: you should use a `struct` insteed of a `class` for such short-lived data. The former is stored on the stack, where as the latter is stored in momory, which is slower to access.

**Comment by unixfox** [comment]
@SamantazFox good to be merged?

**Comment by SamantazFox** [comment]
@unixfox Yep! It's looking good to me!

**Pull request merged**